README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

This Package is used to create a custom steppers inside the Header Message and Header Subtitle Node, It is a of wrapper of flutter_timeline.

## Features

* Easy Customization: Effortlessly customize the colors and alignment in the form of horizontal and vertical to suit your requirements.

* Flexible Size Control: Define the size of the Header Nodes(Message Nodes) according to your needs. Specify the width and height to ensure a perfect fit within your application's UI.

* Use it to show the sports timeline like Football, cricket, order details stepper etc.

* Default it comes with the Header Nodes. you can customise as per your need if Header nodes is not required in your requirements.

## Getting started

import package
```dart
import 'package:flutter_custom_timeline/package.dart';
```

## Usage:-
For Sports like Foot Ball, Cricket, and any other Events Steppers
 ```dart
 void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: FlutterCustomTimeline(
        alignment: TimelineAlign.manual,
        lineXy: 0.3,
        headerXy: -0.7,
        steppers: getStepperData(),
      ),
    );
  }
}
 ```
## Left Align

<img width="330" alt="Screenshot 2023-10-29 at 2 09 52 PM" src="https://github.com/Shadil-rayeen/Flutter-Custom-Timeline/assets/84621229/5d23d60c-dbf5-4279-8642-1b856dc93ed0">

## Center Align
 <img width="333" alt="Screenshot 2023-10-29 at 2 10 24 PM" src="https://github.com/Shadil-rayeen/Flutter-Custom-Timeline/assets/84621229/c043bba7-3da7-4435-a153-74903815b2b7">




## timeLineBuilder-
No  other properties will work expect isHeaderNode if you give your own timeline. Inside you time line Widget properties will work
```agsl
 FlutterCustomTimeline(
        timeLineBuilder: (context, index) {
          return TimelineTile(
            isFirst: true,
            startChild: Text("Date"),
            endChild: Text("Title"),
          );
        },
        //   alignment: TimelineAlign.manual,
        //   lineXy: 0.3,
        //   headerXy: -0.7,
        steppers: getStepperData(),
      )
```

## Additional information

1. headerXy:- It is used to align the header nodes negative double value(-0.7) will align to left and positive value (0.4) will align to right
2. lineXy:- It is ued to align the stepper timeline, Alingment.manual will only work for lineXy
3. steppers:- It will take List of Steppers
4. isFirst:- Used to give the stepper at first
5. isLast:- Used to give the stepper line at last
6. isHeaderNode: bool is used to show the header Node default is 'true'
7. backgroundColor: It used to change the background color of all the steppers
8. afterLineStyle: It is used to style the line after the node
9. beforeLineStyle:- It is used to style the line before node
10. hasIndicator:- bool is used to show the indicator
11. headerStyle: used to style the header Text
12. subHeaderStyle: used to style the sub Header Text
13. headerBackgroundColor:- used to change the header background color
14. timeLineBuilder:- used to full customization your time line.


CHANGELOG.md
## 0.0.6

* Added style for sub nodes


test/flutter_custom_timeline_test.dart
import 'package:flutter_custom_timeline/package.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/widgets/time_line_child.dart
import 'package:flutter/material.dart';

class TimeLineChild extends StatelessWidget {
  const TimeLineChild({
    Key? key,
    this.title,
    this.subtitle,
    this.tittleStyle,
    this.subtitleStyle,
  }) : super(key: key);

  final String? title;
  final String? subtitle;
  final TextStyle? tittleStyle;
  final TextStyle? subtitleStyle;

  @override
  Widget build(BuildContext context) {
    final rowChildren = <Widget>[
      Expanded(
        child: Text(
          title ?? "",
          style: tittleStyle ??
              const TextStyle(
                fontSize: 18,
                fontWeight: FontWeight.bold,
              ),
        ),
      ),
    ];

    return Padding(
      padding: const EdgeInsets.symmetric(vertical: 5),
      child: Column(
        mainAxisSize: MainAxisSize.min,
        mainAxisAlignment: MainAxisAlignment.start,
        crossAxisAlignment: CrossAxisAlignment.start,
        children: <Widget>[
          Row(children: rowChildren),
          Flexible(
            child: Text(
              subtitle ?? "",
              style: subtitleStyle ??
                  const TextStyle(
                    fontSize: 16,
                  ),
            ),
          ),
        ],
      ),
    );
  }
}


lib/widgets/message_timeline.dart
import 'package:flutter/material.dart';

class MessageTimeline extends StatelessWidget {
  const MessageTimeline({
    Key? key,
    required this.message,
    this.subTitle,
    this.backGroundColor,
    this.headerXy,
    this.headerTitleStyle,
    this.headerSubTitleStyle,
  }) : super(key: key);

  final String message;
  final String? subTitle;
  final Color? backGroundColor;
  final double? headerXy;
  final TextStyle? headerTitleStyle;
  final TextStyle? headerSubTitleStyle;

  @override
  Widget build(BuildContext context) {
    return Align(
      alignment: headerXy == null ? Alignment.center : Alignment(headerXy!, headerXy!),
      child: Container(
        padding: const EdgeInsets.all(8),
        margin: const EdgeInsets.symmetric(vertical: 10),
        decoration: BoxDecoration(
          color: backGroundColor ?? Colors.black.withOpacity(0.3),
          borderRadius: const BorderRadius.all(Radius.circular(8)),
        ),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[
            Flexible(
              child: Text(
                message,
                textAlign: TextAlign.center,
                style: headerTitleStyle ?? const TextStyle(fontSize: 16),
              ),
            ),
            if (subTitle != null) ...[
              Flexible(
                child: Text(
                  subTitle!,
                  textAlign: TextAlign.center,
                  style: headerSubTitleStyle ?? const TextStyle(fontSize: 14),
                ),
              ),
            ]
          ],
        ),
      ),
    );
  }
}


lib/screen/flutter_custom_timeline.dart
import 'package:flutter/material.dart';
import 'package:flutter_custom_timeline/widgets/message_timeline.dart';
import 'package:flutter_custom_timeline/widgets/time_line_child.dart';
import 'package:flutter_custom_timeline/widgets/time_line_data.dart';
import 'package:flutter_custom_timeline/widgets/time_line_indocator.dart';
import 'package:timeline_tile/timeline_tile.dart';

class FlutterCustomTimeline extends StatelessWidget {
  const FlutterCustomTimeline({
    super.key,
    required this.steppers,
    this.lineStyle,
    this.timeLineBuilder,
    this.indicatorStyle,
    this.backgroundColor,
    this.extendBodyBehindAppBar = true,
    this.indicatorXY = 0.5,
    this.alignment = TimelineAlign.center,
    this.lineXy,
    this.hasIndicator = true,
    this.afterLineStyle,
    this.isLast = false,
    this.isFirst = false,
    this.headerXy,
    this.headerBackgroundColor,
    this.headerTitleStyle,
    this.headerSubTitleStyle,
    this.isHeaderNode = true,
    this.rightNodeSubtitleStyle,
    this.rightNodeTitleStyle,
    this.leftNodeTitleStyle,
    this.leftNodeSubtitleStyle,
  });

  final List<StepperData> steppers;
  final IndicatorStyle? indicatorStyle;
  final TimelineTile Function(BuildContext context, int index)? timeLineBuilder;
  final LineStyle? lineStyle;
  final Color? backgroundColor;
  final Color? headerBackgroundColor;
  final bool extendBodyBehindAppBar;
  final double indicatorXY;
  final bool hasIndicator;
  final TimelineAlign alignment;
  final double? lineXy;
  final double? headerXy;
  final LineStyle? afterLineStyle;
  final bool isFirst;
  final bool isLast;
  final TextStyle? headerTitleStyle;
  final TextStyle? headerSubTitleStyle;
  final bool isHeaderNode;
  final TextStyle? leftNodeTitleStyle;
  final TextStyle? leftNodeSubtitleStyle;
  final TextStyle? rightNodeTitleStyle;
  final TextStyle? rightNodeSubtitleStyle;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: backgroundColor,
      extendBodyBehindAppBar: extendBodyBehindAppBar,
      body: Column(
        children: [
          Expanded(
            child: ListView.builder(
              shrinkWrap: false,
              itemCount: steppers.length,
              itemBuilder: (context, index) {
                final data = steppers[index].timeLineData;
                return CustomScrollView(
                  physics: const NeverScrollableScrollPhysics(),
                  shrinkWrap: true,
                  slivers: <Widget>[
                    if (steppers[index].headerTitle != null && isHeaderNode) ...[
                      SliverList(
                        delegate: SliverChildListDelegate(
                          <Widget>[
                            MessageTimeline(
                              headerTitleStyle: headerTitleStyle,
                              headerSubTitleStyle: headerSubTitleStyle,
                              message: steppers[index].headerTitle!,
                              headerXy: headerXy,
                              backGroundColor: headerBackgroundColor,
                              subTitle: steppers[index].headerSubtitle,
                            ),
                          ],
                        ),
                      ),
                    ],
                    CustomTimeLine(
                      data: data,
                      timeLineBuilder: timeLineBuilder,
                      indicatorStyle: indicatorStyle,
                      lineStyle: lineStyle,
                      indicatorXY: indicatorXY,
                      alignment: alignment,
                      lineXY: lineXy,
                      hasIndicator: hasIndicator,
                      afterLineStyle: afterLineStyle,
                      isLast: isLast,
                      isFirst: isFirst,
                      leftNodeSubtitleStyle: leftNodeSubtitleStyle,
                      leftNodeTitleStyle: leftNodeTitleStyle,
                      rightNodeSubtitleStyle: rightNodeSubtitleStyle,
                      rightNodeTitleStyle: rightNodeTitleStyle,
                    ),
                  ],
                );
              },
            ),
          ),
        ],
      ),
    );
  }
}

class CustomTimeLine extends StatelessWidget {
  const CustomTimeLine({
    Key? key,
    this.data,
    this.indicatorStyle,
    this.timeLineBuilder,
    this.lineStyle,
    this.isLast = false,
    this.isFirst = false,
    this.hasIndicator = true,
    this.afterLineStyle,
    this.lineXY,
    this.indicatorXY = 0.5,
    this.alignment = TimelineAlign.center,
    required this.rightNodeTitleStyle,
    required this.rightNodeSubtitleStyle,
    required this.leftNodeTitleStyle,
    required this.leftNodeSubtitleStyle,
  }) : super(key: key);

  final List<TimeLineSteps>? data;
  final IndicatorStyle? indicatorStyle;
  final TimelineTile Function(BuildContext context, int index)? timeLineBuilder;
  final LineStyle? lineStyle;
  final LineStyle? afterLineStyle;
  final bool isFirst;
  final bool isLast;
  final bool hasIndicator;
  final double? lineXY;
  final double indicatorXY;
  final TimelineAlign alignment;
  final TextStyle? leftNodeTitleStyle;
  final TextStyle? leftNodeSubtitleStyle;
  final TextStyle? rightNodeTitleStyle;
  final TextStyle? rightNodeSubtitleStyle;

  @override
  Widget build(BuildContext context) {
    return SliverList(
      delegate: SliverChildBuilderDelegate(
        (BuildContext context, int index) {
          final event = data![index];

          final startChild = TimeLineChild(
            title: event.leftTitle,
            subtitle: event.leftSubtitle,
            tittleStyle: leftNodeTitleStyle,
            subtitleStyle: leftNodeSubtitleStyle,
          );
          final endChild = TimeLineChild(
            title: event.rightTitle,
            subtitle: event.rightSubtitle,
            tittleStyle: rightNodeTitleStyle,
            subtitleStyle: rightNodeSubtitleStyle,
          );

          return timeLineBuilder?.call(context, index) ??
              TimelineTile(
                isFirst: isFirst,
                isLast: isLast,
                afterLineStyle: afterLineStyle,
                alignment: alignment,
                startChild: startChild,
                endChild: endChild,
                hasIndicator: hasIndicator,
                lineXY: lineXY,
                indicatorStyle: indicatorStyle ??
                    IndicatorStyle(
                      width: 35,
                      height: 35,
                      indicatorXY: indicatorXY,
                      padding: const EdgeInsets.symmetric(horizontal: 12.0),
                      indicator: TimeLineIndicator(time: "${index + 1}"),
                      drawGap: true,
                    ),
                beforeLineStyle: lineStyle ?? const LineStyle(thickness: 3),
              );
        },
        childCount: data?.length,
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_custom_timeline/package.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: FlutterCustomTimeline(
        alignment: TimelineAlign.manual,
        lineXy: 0.3,
        headerXy: -0.7,
        steppers: getStepperData(),
        rightNodeTitleStyle: const TextStyle(fontSize: 14),
        rightNodeSubtitleStyle: const TextStyle(fontSize: 18),
        leftNodeTitleStyle: const TextStyle(fontSize: 16, color: Colors.red),
      ),
    );
  }
}

List<StepperData> getStepperData() {
  return <StepperData>[
    StepperData(
      headerTitle: "Start of the First Half!!",
      timeLineData: [
        const TimeLineSteps(
          leftTitle: "4:30 AM",
          rightTitle: 'Foul!',
          rightSubtitle: 'David Luiz brings his opponent down.',
        ),
        const TimeLineSteps(
          rightTitle: 'Yellow',
          leftTitle: "4:30 AM",
          rightSubtitle: 'This yellow card was deserved.',
        ),
        const TimeLineSteps(
          leftTitle: "4:30 AM",
          rightTitle: 'Gooooaaaal!',
          rightSubtitle: 'Goal! Lionel Messi slams the ball into the open net from close range.',
        ),
        const TimeLineSteps(
          rightTitle: 'One more!',
          leftTitle: "4:30 AM",
          rightSubtitle: 'Piqué gets a yellow card for arguing with the referee.',
        ),
        const TimeLineSteps(
          leftTitle: "4:30 AM",
          rightTitle: 'Ouchh',
          rightSubtitle: 'Mesut Özil is writhing in pain and can now receive medical treatment.',
        ),
      ],
    ),
    StepperData(
      headerTitle: "Start of the Second Half!!",
      timeLineData: [
        const TimeLineSteps(
          leftTitle: "4:30 AM",
          rightTitle: 'Foul!',
          rightSubtitle: 'David Luiz brings his opponent down.',
        ),
        const TimeLineSteps(
          rightTitle: 'Yellow',
          leftTitle: "4:30 AM",
          rightSubtitle: 'This yellow card was deserved.',
        ),
        const TimeLineSteps(
          leftTitle: "4:30 AM",
          rightTitle: 'Gooooaaaal!',
          rightSubtitle: 'Goal! Lionel Messi slams the ball into the open net from close range.',
        ),
        const TimeLineSteps(
          rightTitle: 'One more!',
          leftTitle: "4:30 AM",
          rightSubtitle: 'Piqué gets a yellow card for arguing with the referee.',
        ),
        const TimeLineSteps(
          leftTitle: "4:30 AM",
          rightTitle: 'Ouchh',
          rightSubtitle: 'Mesut Özil is writhing in pain and can now receive medical treatment.',
        ),
      ],
    ),
  ];
}
